bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+package org.mozartoz.bootcompiler
+package transform
+
+import ast._
+import oz._
+import symtab._
+import scala.collection._
+import com.sun.javafx.fxml.expression.BinaryExpression
+
+sealed class VarState(val onStack: Boolean, val locked: Boolean) {
+}
+object Declared      extends VarState(false, false)
+object Captured      extends VarState(false, false)
+object BoundFirst    extends VarState(true, true)
+object ToInstantiate extends VarState(false, true)
+
+case class VarAndState(v: Variable, state: VarState)
+
+/**
+ * Looks for declarations and the usages of those declarations.
+ * The Transform methods detect local statements and add declared variables to the environment
+ * The Walk methods are collecting their usages and tagging them
+ * 
+ * If a variable is bound after its declaration, we can shortcut it
+ * 
+ * When a proc is declared, it is difficult to track where it will end up. So if a call occurs,
+ * by precaution, all captured variable that were not bound should be instantiated
+ * 	This means that any A = {IntToFloat 3} will instantiate because it results in {IntToFloat 3 A}
+ *  that could potentially call end up calling A
```

Comment:
The idea here was more that A could be a procedure, so already bound, used as parameter to be called later (The IntToFloat was a bad example however, it would more be a custom map which would take the function as last argument)